src/wasm/wasm-validator.cpp

```diff
@@ -406,6 +407,44 @@ struct FunctionValidator : public WalkerPass<PostWalker<FunctionValidator>> {
     size_t align, Type type, Index bytes, bool isAtomic, Expression* curr);
   void validateMemBytes(uint8_t bytes, Type type, Expression* curr);
 
+  template<typename T>
+  void validateCallParamsAndResult(T* curr, Signature sig) {
+    if (!shouldBeTrue(curr->operands.size() == sig.params.size(),
+                      curr,
+                      "call* param number must match")) {
+      return;
+    }
+    size_t i = 0;
+    for (const auto& param : sig.params) {
+      if (!shouldBeSubTypeOrFirstIsUnreachable(curr->operands[i]->type,
+                                               param,
+                                               curr,
+                                               "call param types must match") &&
+          !info.quiet) {
+        getStream() << "(on argument " << i << ")\n";
+      }
+      ++i;
+    }
+    if (curr->isReturn) {
+      shouldBeEqual(curr->type,
+                    Type(Type::unreachable),
+                    curr,
+                    "return_call* should have unreachable type");
+      shouldBeEqual(
+        getFunction()->sig.results,
+        sig.results,
+        curr,
+        "return_call* callee return type must match caller return type");
+    } else {
+      if (curr->type != Type::unreachable) {
+        shouldBeEqual(curr->type,
```

Comment:
This looks like an opportunity to use `shouldBeEqualOrFirstIsUnreachable`.

Reply:
Sounds good, improving.